protocol/consistencychecks.go

```diff
@@ -0,0 +1,298 @@
+// state of a CONIKS client. This includes the current epoch
+// and SignedTreeRoot value, as well as directory's policies
+// (e.g., wherether the TemporaryBinding extension is being used or not)
+// The client should create a new ConsistencyChecks instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConsistencyChecks instance then will be used to verify
+// the returned responses from the ConiksDirectory.
+type ConsistencyChecks struct {
+	CurrentEpoch uint64
+	SavedSTR     []byte
+
+	// extensions settings
+	useTBs bool
+	TBs    map[string]*m.TemporaryBinding
+
+	// signing key
+	// TODO: should we pin the signing key in the client? see #47
+	signKey sign.PublicKey
+}
+
+// NewCC creates an instance of ConsistencyChecks using
+// the pinning directory's STR at epoch 0.
+func NewCC(savedSTR []byte, useTBs bool, signKey sign.PublicKey) *ConsistencyChecks {
+
+	// TODO: see #110
+	if !useTBs {
+		panic("Currently the server is forced to use TBs")
+	}
+
+	cc := &ConsistencyChecks{
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		useTBs:       useTBs,
+		signKey:      signKey,
+	}
+
+	if useTBs {
+		cc.TBs = make(map[string]*m.TemporaryBinding)
+	}
+	return cc
+}
+
+// Verify verifies the directory's response for a request.
+// It first verifies the directory's returned status code of the request.
+// If the status code is not in ErrorResponses array, it means
+// the directory has successfully handled the request.
+// The verifier will then verify the consistency state of the response,
+// and finally, it verifies the returned proof type
+// according to the request type.
+// This will also update the consistency state if the check is passed.
+func (cc *ConsistencyChecks) Verify(requestType int, msg *Response,
+	uname string, key []byte) ErrorCode {
+
+	if ErrorResponses[msg.Error] {
+		return msg.Error
+	}
+
+	var verifResult error
+	var proofType int
+
+	switch msg.DirectoryResponse.(type) {
+	case *DirectoryProof:
+		proofType, verifResult = cc.verifyDirectoryProof(
+			msg.DirectoryResponse.(*DirectoryProof), uname, key)
+	case *DirectoryProofs:
+		proofType, verifResult = cc.verifyDirectoryProofs(
+			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	}
+
+	if verifResult != nil {
+		return verifResult.(ErrorCode)
+	}
+
+	switch requestType {
+	case RegistrationType:
+		if msg.Error == ErrorNameExisted &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		} else if msg.Error == Success &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
+
+	case KeyLookupType:
+		if msg.Error == ErrorNameNotFound &&
+			proofType == proofOfAbsence {
+			return PassedWithAProofOfAbsence
+		} else if msg.Error == Success &&
+			proofType == proofOfInclusion {
+			return PassedWithAProofOfInclusion
+		}
+		return cc.verifyProofTypeUsingPromises(msg.Error, proofType)
```

Comment:
Shouldn't reuse this method here.  `e == ErrorNameExisted && proofType == proofOfAbsence` doesn't apply.